internal/migrate/staging_validator.go

```diff
@@ -0,0 +1,427 @@
+/*
+ * Flow CLI
+ *
+ * Copyright 2019 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package migrate
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/onflow/flow-cli/internal/util"
+
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/runtime"
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/onflow/cadence/runtime/old_parser"
+	"github.com/onflow/cadence/runtime/parser"
+	"github.com/onflow/cadence/runtime/pretty"
+	"github.com/onflow/cadence/runtime/sema"
+	"github.com/onflow/cadence/runtime/stdlib"
+	"github.com/onflow/contract-updater/lib/go/templates"
+	flowsdk "github.com/onflow/flow-go-sdk"
+	"github.com/onflow/flow-go/cmd/util/ledger/migrations"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flowkit/v2"
+)
+
+type stagingValidator struct {
+	flow  flowkit.Services
+	state *flowkit.State
+
+	// Cache for account contract names so we don't have to fetch them multiple times
+	accountContractNames map[common.Address][]string
+	// All resolved contract code
+	contracts map[common.Location][]byte
+	// Record errors related to missing staged dependencies, as these are reported separately
+	missingDependencies []common.AddressLocation
+	// Cache for contract elaborations which are reused during program checking & used for the update checker
+	elaborations map[common.Location]*sema.Elaboration
+}
+
+type accountContractNamesProviderImpl struct {
+	resolverFunc func(address common.Address) ([]string, error)
+}
+
+var _ stdlib.AccountContractNamesProvider = &accountContractNamesProviderImpl{}
+
+type missingDependenciesError struct {
+	MissingContracts []common.AddressLocation
+}
+
+func (e *missingDependenciesError) Error() string {
+	contractNames := make([]string, len(e.MissingContracts))
+	for i, location := range e.MissingContracts {
+		contractNames[i] = location.Name
+	}
+	return fmt.Sprintf("the following staged contract dependencies could not be found (have they been staged yet?): %v", contractNames)
+}
+
+var _ error = &missingDependenciesError{}
+
+var chainIdMap = map[string]flow.ChainID{
```

Comment:
I've had to do this several times as well. I wonder if there is a way we could break this out to be reusable.

Reply:
Yeah, just a flowkit helper for converting `flowkit` chain id to flow-go chainid would be good enough here